### ISSUE TITLE
Cancelling tests do not work properly and spuriously fails

### ISSUE THREAD

**Issue by StephanEwen** [stratosphere/stratosphere#154]
The cancelling tests do not actually report errors while cancelling and may hang.

**Comment by rmetzger** [comment]
(This is a little brain dump from yesterdays meeting with Stephan. I hope I remember everything correctly)
The idea of the cancelling tests is to verify that failed tasks are properly shut down.

The current implementation of the testcase is a bit wrong. 
There is a main thread (a) that starts another thread (b) which is going to be cancelled eventually. (a) waits until (b) has been cancelled.. 
But there are junit-checks in the code of thread (b) which sometimes cause the thread (b) to die. (a) then waits indefinitely for (b) to stop, which causes stalling tests as reported in https://github.com/dimalabs/ozone/issues/94.
What needs to be done is to somehow report test errors to thread (a), so that junit properly recognizes the failure.

**Comment by StephanEwen** [comment]
There is a thread (c) that calls cancel on the tasks. (a) then waits for (b) and (c) to finish.

When the cancelling logic is faulty itself, (c) will terminate without letting the testcase know that something went wrong.

**Comment by rmetzger** [comment]
Today, I got this NPE

```
14:19:08,130 INFO  eu.stratosphere.nephele.execution.ExecutionStateTransition    - TM: ExecutionState set from CANCELING to CANCELED for task Match on string key (6/64)
14:19:08,134 ERROR eu.stratosphere.nephele.taskmanager.runtime.RuntimeTask       - java.lang.NullPointerException
        at eu.stratosphere.nephele.services.memorymanager.spi.DefaultMemoryManager.release(DefaultMemoryManager.java:337)
        at eu.stratosphere.pact.runtime.hash.BuildFirstHashMatchIterator.abort(BuildFirstHashMatchIterator.java:171)
        at eu.stratosphere.pact.runtime.task.MatchDriver.cancel(MatchDriver.java:185)
        at eu.stratosphere.pact.runtime.task.RegularPactTask.cancel(RegularPactTask.java:416)
        at eu.stratosphere.nephele.taskmanager.runtime.RuntimeTask.cancelOrKillExecution(RuntimeTask.java:227)
        at eu.stratosphere.nephele.taskmanager.runtime.RuntimeTask.cancelExecution(RuntimeTask.java:186)
        at eu.stratosphere.nephele.taskmanager.TaskManager$1.run(TaskManager.java:399)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1146)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:679)

14:19:08,136 WARN  eu.stratosphere.pact.runtime.task.RegularPactTask             - PACT code cancelled.: Match on string key (13/64)
```

It occurred on a failed task.

The `availableMemory` of the HashMap seem to contain `null` memory segments.

@StephanEwen: Should I open a new issue for this or is it related to this issue?

**Comment by StephanEwen** [comment]
Open an issue, please...
Am 22.11.2013 14:23 schrieb "Robert Metzger" notifications@github.com:

**Comment by StephanEwen** [comment]
Fixed in 7b0c53b4c501690866523468f6c3e7f06438658c
